rhp/v2/contracts.go

```diff
@@ -69,52 +69,49 @@ func ContractRenewalCost(fc types.FileContract, contractFee types.Currency) type
-		timeExtension := uint64(contractEnd - currentRevision.WindowEnd)
-		basePrice = host.StoragePrice.Mul64(currentRevision.Filesize).Mul64(timeExtension)
-		baseCollateral = host.Collateral.Mul64(currentRevision.Filesize).Mul64(timeExtension)
-	}
 
-	// estimate collateral for new contract
-	var newCollateral types.Currency
-	if costPerByte := host.UploadBandwidthPrice.Add(host.StoragePrice).Add(host.DownloadBandwidthPrice); !costPerByte.IsZero() {
-		bytes := renterPayout.Div(costPerByte)
-		newCollateral = host.Collateral.Mul(bytes)
+	// if the contract height did not increase both prices are zero
+	if contractEnd := uint64(endHeight + s.WindowSize); contractEnd > fc.WindowEnd {
+		timeExtension := uint64(contractEnd - fc.WindowEnd)
+		basePrice = s.StoragePrice.Mul64(fc.Filesize).Mul64(timeExtension)
+		baseCollateral = s.Collateral.Mul64(fc.Filesize).Mul64(timeExtension)
 	}
 
-	// the collateral can't be greater than MaxCollateral
-	totalCollateral := baseCollateral.Add(newCollateral)
-	if totalCollateral.Cmp(host.MaxCollateral) > 0 {
-		totalCollateral = host.MaxCollateral
-	}
-
-	// Calculate payouts: the host gets their contract fee, plus the cost of the
-	// data already in the contract, plus their collateral. In the event of a
-	// missed payout, the cost and collateral of the data already in the
-	// contract is subtracted from the host, and sent to the void instead.
-	//
-	// However, it is possible for this subtraction to underflow; this can
-	// happen if baseCollateral is large and MaxCollateral is small. We cannot
-	// simply replace the underflow with a zero, because the host performs the
-	// same subtraction and returns an error on underflow. Nor can we increase
-	// the valid payout, because the host calculates its collateral contribution
-	// by subtracting the contract price and base price from this payout, and
-	// we're already at MaxCollateral. Thus the host has conflicting
-	// requirements, and renewing the contract is impossible until they change
-	// their settings.
-	hostValidPayout := host.ContractPrice.Add(basePrice).Add(totalCollateral)
+	hostValidPayout := s.ContractPrice.Add(basePrice).Add(hostCollateral)
 	voidMissedPayout := basePrice.Add(baseCollateral)
 	if hostValidPayout.Cmp(voidMissedPayout) < 0 {
 		// TODO: detect this elsewhere
 		panic("host's settings are unsatisfiable")
 	}
 	hostMissedPayout := hostValidPayout.Sub(voidMissedPayout)
+	return hostValidPayout, hostMissedPayout, voidMissedPayout
+}
+
+// PrepareContractRenewal constructs a contract renewal transaction.
+func PrepareContractRenewal(currentRevision types.FileContractRevision, renterKey types.PrivateKey, hostKey types.PublicKey, renterPayout, hostCollateral types.Currency, endHeight uint64, host HostSettings, refundAddr types.Address) types.FileContract {
+	// the collateral can't be greater than MaxCollateral
+	if hostCollateral.Cmp(host.MaxCollateral) > 0 {
```

Comment:
At this point the `hostCollateral` doesn't include the `baseCollateral` yet. So it might still exceed the `MaxCollateral`.

Reply:
it does

Reply:
I'll update the PR to pass in base price, base collateral and host collateral so we are providing all info.